### PR TITLE
Versioning

### PR DESCRIPTION
**Pull request by JonathonDunford**
Duplicate of #26, but separated.

VERSIONING!

Migration of funds with one request through the smart contract.
Successor / predecessor contracts.
Version # storage.

Big thanks to TokenStore for providing the idea and a decent bit of the implementation: https://github.com/tokenstore/contract/blob/master/contracts/TokenStore.sol

+Some improved in-code documentation.

### CODE REVIEW

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Should be explicitly defined as public.
newContract variable is only used when emitting the event at the end, but not used as the target!
newContract should not be passed at all, imho, but, just as the code does it, successor that was set on creation by the admin/contact creator, should be used. 
I generally refer the way it is done in the TokenStore contract (searching the latest contract in a limited loop) - one thing less the user can mess up and also a better security footprint.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Line can go. Function is not defined as payable.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: 0 for Ether is NOT allowed in this case.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: +1 on removing the pseudo-check "require(caller.version() > 0);"
I can't spot a problem here, but we will need a number of test cases for this... I have a feeling of opening the gates here.

**Review by mmhh1910 — CHANGES_REQUESTED** [review]
migrateFunds needs fixing before approval.
Other proposed changes as you see fit.

File: contracts/ForkDelta.sol
Comment (JonathonDunford) [reply contracts/ForkDelta.sol]: My reasoning behind passing in the contract was because:
- It simplifies the logic in the contract
- It allows us / users to migrate funds to any ForkDelta contract (or really any contract that makes use of our protocol)

I am open to suggestions, but it just seems like one less problem if we can select the specific contract to "upgrade" to especially if we have multiple. I understand the looped logic did that, but only to an extent and it was somewhat cumbersome.

**Comment by JonathonDunford** [comment]
@betterdelta Changes have been made.

**Comment by mmhh1910** [comment]
We now have a public successor variable that is never set and a predecessor variable that is set on contract creation, but never actually used.

**Comment by JonathonDunford** [comment]
@betterdelta 

How do you recommend fixing that?

Adding a function usable by admin to set successor?

I was thinking those variables would be moreso used to keep track of the versioning on the actual blockchain, so you can see previous contracts and find the most up to date.

**Comment by mmhh1910** [comment]
> How do you recommend fixing that?

First, decide if you want to a.) give the target contract on migrateFunds as a parameter of the function (the way you did it) or b.) you want to do it the TokenStore way (looping through the successors, which are set by the admin once the contract is deprecated).

If a.) I'd remove predecessors and successor alltogether. 

If b.) there is some functionality missing from the TokenStore contract like 

```
function deprecate(bool _deprecated, address _successor) onlyOwner {
    deprecated = _deprecated;
    successor = _successor;
  }
```

**Comment by JonathonDunford** [comment]
@betterdelta 

If a) you don't think it's important for the contract to store where the new or previous versions are?

Side note: The deprecated flag is kind of unnecessary. If successor is address(0), it hasn't been deprecated.

**Comment by mmhh1910** [comment]
> If a) you don't think it's important for the contract to store where the new or previous versions are?

I personally would not store this kind of information in the contract if it's not functional needed.

**Comment by JonathonDunford** [comment]
@betterdelta 

Can you give a reason as to why?

It seems important information for a user that is viewing the smart contract to see if it is the current one (or which one is)

**Comment by mmhh1910** [comment]
I think users expect to see this in the UI, not the contract.

**Comment by JonathonDunford** [comment]
@betterdelta 

Of course, but are there any benefits to excluding this information from the smart contract?
This would allow users on etherscan (or any other service) to see this information. It would also allow bots to automatically update to the most up to date contract by watching the `successor` variable.

**Comment by mmhh1910** [comment]
This kind of automation would be nice, which is why I would go with b.). 

I'm the wrong person to ask how to fix a.). ;-)

**Comment by JonathonDunford** [comment]
@betterdelta 

If there aren't any benefits to excluding this information from the smart contract, I think it is safe to say we should keep it included.

**Comment by JonathonDunford** [comment]
@betterdelta 

Added a function for the admin to set the successor value. Good catch.

Another reason I prefer this method is that the contract doesn't have implicit control over where the user chooses to migrate their funds.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Can we name this in a more describing way? Like updateSuccessor or setSuccessor. 

And just out of habit add a: require(successor != address(0)); ?

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Nitpicking: I'd prefer the code style:

```
if (tokenAmount != 0) {
      require(Token(token).approve(newExchange, tokenAmount));
      tokens[token][msg.sender] = 0;
      newExchange.depositTokenForUser(token, tokenAmount, msg.sender);
}
```

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: If it is only intended for use from migrateFunds, we could declare the two depositForUser functions private. That would also lower the attack surface.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: Same as above.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: You still have to clean up this merge.

File: contracts/ForkDelta.sol
Comment (mmhh1910) [inline contracts/ForkDelta.sol]: n is not declared:

`for (uint16 n = 0; n < tokens_.length; n++) {`

**Review by mmhh1910 — CHANGES_REQUESTED** [review]
>Another reason I prefer this method is that the contract doesn't have implicit control over where the 
> user chooses to migrate their funds.

Actually, the contract having control where funds should be send to is why I prefer b.). (Lower attack surface.)

The code still needs a few changes and currently does not compile.

File: contracts/ForkDelta.sol
Comment (JonathonDunford) [reply contracts/ForkDelta.sol]: I agree.

**Comment by JonathonDunford** [comment]
@betterdelta 

That opens this topic up for discussion:

Do we actually want to make the "deposit for user" functions private or is there some use case in the future where we may want to keep them public?

I could see that being useful in:
- gifting
- some sort of trading competitions from token issuers (encouraging users to trade even more)

Cons:
- Introduces 2 attack vectors

Although, we will be having a security audit, so that shouldn't be an issue.

However, I'm not entirely sold either way. What are your thoughts for/against?

**Comment by JonathonDunford** [comment]
depositFor functions need to be public so they can be accessed from the separate contract.

**Review by mmhh1910 — APPROVED** [review]

**Pull request merged**
